File: practice/src/main/java/com/decade/practice/persistence/jpa/entities/SyncContext.java
package com.decade.practice.persistence.jpa.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.MapsId;
import jakarta.persistence.OneToOne;
import lombok.Data;

import java.util.UUID;

@Entity
@Data
public class SyncContext {
    public static final int STARTING_VERSION = 0;

    @MapsId
    @OneToOne
    @JsonIgnore
    private User owner;

    @Id
    private UUID id;

    private int eventVersion = STARTING_VERSION;

    // No-arg constructor required by JPA
    protected SyncContext() {
    }

    public SyncContext(User owner) {
        this.owner = owner;
    }

    public void incVersion() {
        this.setEventVersion(this.getEventVersion() + 1);
    }

}

File: practice/src/main/java/com/decade/practice/model/SyncContext.kt
package com.decade.practice.model

import com.decade.practice.model.entity.User
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.MapsId
import jakarta.persistence.OneToOne
import java.util.*


const val STARTING_VERSION = 0

@Entity
data class SyncContext(
    // primary key is derived
    @MapsId // PrimaryKeyJoinColumn
    @OneToOne
    var owner: User
) {
    @Id
    var id: UUID? = null

    var eventVersion: Int = STARTING_VERSION

    fun incVersion(): Int {
        return ++this.eventVersion
    }
}
